File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;z
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isSameTree(TreeNode *root, TreeNode *subRoot){
        if (!root && ! subRoot) return true;
        else if (!root || !subRoot) return false;
        if(root->val != subRoot-> val) return false;
        return isSameTree(root->left, subRoot->left) && isSameTree(root->right, subRoot->right);
            }
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(!root) return false;
        if(isSameTree(root, subRoot)) return true;
        if(isSubtree(root->right, subRoot) || isSubtree(root->left, subRoot)) return true;
        return false;
        
        

    }
};

File: Data Structures & Algorithms/house-robber-iii/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int [] dfs(TreeNode root){
        if (root == null) return new int[] {0,0};
        int val = root.val;

        int[] l =  dfs(root.left);
        int[] r =  dfs(root.right);
        
        int rob =  (val + l[0] + r[0]);

    int notrob =  Math.max (l[0], l[1]) + Math.max (r[0], r[1]);

    return new int[] {notrob, rob};
        
            }
    public int rob(TreeNode root) {
        int [] res = dfs(root);
        return Math.max(res[0], res[1]);
    }
}